Source/Fuse.Controls.Native/iOS/InputDispatch.uno

```diff
@@ -8,131 +8,197 @@ namespace Fuse.Controls.Native.iOS
+			RaiseEvent(EventType.Pressed, origin, touches);
+		}
 
-		static InputDispatch()
+		public static void OnTouchesMoved(Visual origin, ObjC.Object touches)
 		{
-			_listeners = new Dictionary<ObjC.Object, Visual>();
-			_activeTouches = new List<UITouch>();
-			_eventHandler = CreateEventHandler(OnTouchEvent);
+			RaiseEvent(EventType.Moved, origin, touches);
 		}
 
-		static void OnTouchEvent(ObjC.Object sender, ObjC.Object uiEvent)
+		public static void OnTouchesEnded(Visual origin, ObjC.Object touches)
 		{
-			if (sender.IsUIControl() && uiEvent.IsUIEvent())
-			{
-				if (_listeners.ContainsKey(sender))
-				{
-					var visual = _listeners[sender];
-					var ev = new UIEvent(uiEvent);
-					HandleEvent(sender, visual, ev);
-				}
-			}
+			RaiseEvent(EventType.Released, origin, touches);
 		}
 
-		static void ActivateTouch(UITouch touch)
+		public static void OnTouchesCancelled(Visual origin, ObjC.Object touches)
 		{
-			if (!_activeTouches.Contains(touch))
-				_activeTouches.Add(touch);
+			RaiseEvent(EventType.Cancelled, origin, touches);
 		}
 
-		static int GetPointerIndex(UITouch touch)
+		[Foreign(Language.ObjC)]
+		public static void AddInputHandler(Visual owner, ViewHandle viewHandle)
+		@{
+			UIView* view = (UIView*)@{Fuse.Controls.Native.ViewHandle:Of(viewHandle).HitTestHandle:Get()};
+			addInputHandler(view, ^void(int type, id<UnoObject> visual, id touches) { @{Fuse.Controls.Native.iOS.InputDispatch.RaiseEvent(int,object,ObjC.Object):Call(type,visual,touches)}; }, owner);
+		@}
+
+		[Foreign(Language.ObjC)]
+		public static void RemoveInputHandler(ViewHandle viewHandle)
+		@{
+			UIView* view = (UIView*)@{Fuse.Controls.Native.ViewHandle:Of(viewHandle).HitTestHandle:Get()};
+			removeInputHandler(view);
+		@}
+
+		static void RaiseEvent(int eventType, object origin, ObjC.Object touches)
 		{
-			return _activeTouches.IndexOf(touch);
+			RaiseEvent((EventType)eventType, (Visual)origin, touches);
 		}
 
-		static void DeactivateAllTouches()
+		enum EventType : int
 		{
-			_activeTouches.Clear();
+			Pressed = 0,
+			Moved = 1,
+			Released = 2,
+			Cancelled = 3,
 		}
 
-		public static void HandleEvent(ObjC.Object viewHandle, Visual origin, ObjC.Object uiEvent)
+		static readonly Action<Visual,PointerEventData>[] _handlers = new Action<Visual,PointerEventData>[] {
+			RaisePressed,
+			RaiseMoved,
+			RaiseReleased,
+			RaiseCancelled,
+		};
+
+		static void RaiseEvent(EventType type, Visual origin, ObjC.Object touches)
 		{
-			if (uiEvent.IsUIEvent())
-				HandleEvent(viewHandle, origin, new UIEvent(uiEvent));
+			var rootInfo = GetRootInfo(origin);
+			var count = NSArrayCount(touches);
+			var handler = _handlers[(int)type];
+			var deactivate = ((int)type) > ((int)EventType.Moved);
+			for (var i = 0; i < count; i++)
+			{
+				var uiTouch = NSArrayObjectAtIndex(touches, i);
+				handler(rootInfo.RootVisual, NewPointerEventData(uiTouch, rootInfo.RootView));
+				if (deactivate)
+					DeactivateTouch(uiTouch);
+			}
 		}
```

Comment:
its to avoid lots of duplicated code in both C# and Objective-C. Also to avoid allocating 4 delegates for every native view.